dev/io.openliberty.io.smallrye.config/src/io/openliberty/microprofile/config/internal/extension/ConfigWrapper.java

```diff
@@ -37,8 +41,14 @@ void addApplication(String appName) {
     boolean removeApplication(String appName) {
         boolean close = false;
         boolean removed = this.applications.remove(appName);
-        if (removed && (this.applications.size() == 0)) {
-            close = true;
+        if (removed) {
+            //if all applications except for the global one have been removed, remove that too
+            if (this.applications.size() == 1) {
+                this.applications.remove(GLOBAL_CONFIG_APPLICATION_NAME);
+            }
+            if (this.applications.size() == 0) {
+                close = true;
+            }
```

Comment:
I'm unsure of this logic. It would be quite possible for a liberty feature to request a config and to still need it even if all applications are removed.

Then again, is there anything that would stop them using the config after this point if we close it? Usually using a resource after it's been closed causes an exception, but I'm not sure if that's the case here.

Reply:
what is the useful of it when all apps are closed?

Reply:
Unlike our previous implementations, all this does is to remove the config from caches. It does not actually close them. Therefore, if an internal feature still has a reference then it should still work. If it does not then a new one should be created upon request.

Reply:
> what is the useful of it when all apps are closed?

There are lots of features which do things even if no apps are deployed. E.g. MP Metrics will still report the JVM and app server metrics, MP OpenAPI will report an empty OpenAPI document. I'm more concerned about whether they'll cause errors rather than if their functionality with no applications is useful.

Reply:
> Unlike our previous implementations, all this does is to remove the config from caches. It does not actually close them.

It's not clear that that's the case from this bit of code. Possibly renaming the `close` variable and adding some javadoc to the method would help.

Reply:
yeah, fair enough, `close` was a hangover from the previous version

Reply:
> > what is the useful of it when all apps are closed?
> 
> There are lots of features which do things even if no apps are deployed. E.g. MP Metrics will still report the JVM and app server metrics, MP OpenAPI will report an empty OpenAPI document. I'm more concerned about whether they'll cause errors rather than if their functionality with no applications is useful.

It closes `config`. I am not sure MP Metrics and MP OpenAPI needs a runtime config laying around in the background, besides MP Metrics has no dependency on config.

Reply:
Technically it does not "close" the config ... the standard Config API does not have a close method. All it does is call `releaseConfig(Config)` which SmallRye implements as simply removing the Config from the cache.
https://github.com/smallrye/smallrye-config/blob/08918436d1a48b411da530be99e3da97df1e5bfe/implementation/src/main/java/io/smallrye/config/SmallRyeConfigProviderResolver.java#L124